flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/common/matrix/Vector.java

```diff
@@ -0,0 +1,298 @@
+		return SparseVector.deserialize(str);
+	}
+
+	public static Tuple2 <int[], float[]> parseSparseTensor(String str) {
+		int numValues = 1;
+		for (int i = 0; i < str.length(); i++) {
+			if (str.charAt(i) == ',') {
+				numValues++;
+			}
+		}
+		int[] indices = new int[numValues];
+		float[] values = new float[numValues];
+
+		int startPos = StringUtils.lastIndexOf(str, '$') + 1;
+		int endPos = -1;
+		int delimiterPos;
+
+		for (int i = 0; i < numValues; i++) {
+			// extract the value string
+			endPos = StringUtils.indexOf(str, ',', startPos);
+			if (endPos == -1) {
+				endPos = str.length();
+			}
+			delimiterPos = StringUtils.indexOf(str, ':', startPos);
+			if (delimiterPos == -1) {
+				throw new RuntimeException("invalid data: " + str);
+			}
+			indices[i] = Integer.valueOf(StringUtils.substring(str, startPos, delimiterPos));
+			values[i] = Float.valueOf(StringUtils.substring(str, delimiterPos + 1, endPos));
+			startPos = endPos + 1;
+		}
+
+		return Tuple2.of(indices, values);
+	}
+
+	public static boolean isSparse(String str) {
+		if (org.apache.flink.util.StringUtils.isNullOrWhitespaceOnly(str)) {
+			return true;
+		}
+		return StringUtils.indexOf(str, ':') != -1 || StringUtils.indexOf(str, "$") != -1;
+	}
+
+	public static Vector deserialize(String str) {
+		Vector vec;
+		if (isSparse(str)) {
+			vec = Vector.sparse(str);
+		} else {
+			vec = Vector.dense(str);
+		}
+		return vec;
+	}
+
+	public static Vector plus(Vector vec1, Vector vec2) {
+		if (vec1 instanceof DenseVector && vec2 instanceof DenseVector) {
+			return ((DenseVector) vec1).plus((DenseVector) vec2);
+		} else if (vec1 instanceof SparseVector && vec2 instanceof SparseVector) {
+			return ((SparseVector) vec1).plus((SparseVector) vec2);
+		} else if (vec1 instanceof SparseVector && vec2 instanceof DenseVector) {
+			return ((SparseVector) vec1).plus((DenseVector) vec2);
+		} else if (vec1 instanceof DenseVector && vec2 instanceof SparseVector) {
+			return ((SparseVector) vec2).plus((DenseVector) vec1);
+		} else {
+			throw new RuntimeException("Not implemented yet!");
+		}
+	}
+
+	public static Vector minus(Vector vec1, Vector vec2) {
+		if (vec1 instanceof DenseVector && vec2 instanceof DenseVector) {
+			return ((DenseVector) vec1).minus((DenseVector) vec2);
+		} else if (vec1 instanceof SparseVector && vec2 instanceof SparseVector) {
+			return ((SparseVector) vec1).minus((SparseVector) vec2);
+		} else if (vec1 instanceof SparseVector && vec2 instanceof DenseVector) {
+			return ((SparseVector) vec1).minus((DenseVector) vec2);
+		} else if (vec1 instanceof DenseVector && vec2 instanceof SparseVector) {
+			return ((SparseVector) vec2).scale(-1.0).plus((DenseVector) vec1);
+		} else {
+			throw new RuntimeException("Not implemented yet!");
+		}
+	}
+
+	public static double dot(Vector vec1, Vector vec2) {
+		if (vec1 instanceof DenseVector && vec2 instanceof DenseVector) {
+			return ((DenseVector) vec1).dot((DenseVector) vec2);
```

Comment:
Can plus, minus, dot and some other methods be added to the Vector interface? The subclasses override these methods and determine what to do with different type of input vector. Users can then easily use vec1.dot(vec2) to get the result rather than using the method, or the method can be reserved with the simple function call. The only different with current is that the return type of plus and minus in the subclasses will be Vector rather than DenseVector or SparseVector.

Reply:
Thanks for your suggestion, we refactored the implementation.